plugin/pkg/auth/authenticator/token/openid/openid.go

```diff
@@ -0,0 +1,156 @@
+
+type OpenIDAuthenticator struct {
+	client *oidc.Client
+}
+
+// NewOpenID creates a new OpenID client with the given clientID and discovery URL.
+// NOTE(yifan): For now we assume the server provides the "jwks_uri" so we don't
+// need to manager the key sets by ourselves.
+func NewOpenID(clientID, discovery, caFile string) (*OpenIDAuthenticator, error) {
+	var cfg oidc.ProviderConfig
+	var err error
+	var roots *x509.CertPool
+
+	url, err := url.Parse(discovery)
+	if err != nil {
+		return nil, err
+	}
+
+	if url.Scheme == "http" {
+		glog.Warning("Using insecure connection to the OpenID service")
+	} else {
+		if caFile != "" {
+			roots, err = util.CertPoolFromFile(caFile)
+			if err != nil {
+				glog.Errorf("Failed to read the CA file: %v", err)
+			}
+		}
+		if roots == nil {
+			glog.Info("No x509 certificates provided, will use system root CA files")
+		}
+	}
+
+	// Copied from http.DefaultTransport.
+	tr := &http.Transport{
+		TLSClientConfig: &tls.Config{RootCAs: roots},
+		Proxy:           http.ProxyFromEnvironment,
+		Dial: (&net.Dialer{
+			Timeout:   30 * time.Second,
+			KeepAlive: 30 * time.Second,
+		}).Dial,
+		TLSHandshakeTimeout: 10 * time.Second,
+	}
+
+	hc := &http.Client{}
+	hc.Transport = tr
+
+	for i := 0; i <= maxRetries; i++ {
+		if i == maxRetries {
+			return nil, fmt.Errorf("Failed to fetch provider config after %v retries", maxRetries)
+		}
+
+		cfg, err = oidc.FetchProviderConfig(hc, discovery)
+		if err == nil {
+			break
+		}
+		glog.Errorf("Failed to fetch provider config, trying again in %v: %v", retryBackoff, err)
+		time.Sleep(retryBackoff)
+	}
+
+	glog.Infof("Fetched provider config from %s: %#v", discovery, cfg)
+
+	if cfg.KeysEndpoint == "" {
+		return nil, errNoKeyEndpoint
+	}
+
+	ccfg := oidc.ClientConfig{
+		HTTPClient:     hc,
+		Credentials:    oidc.ClientCredentials{ID: clientID},
+		ProviderConfig: cfg,
+	}
+
+	client, err := oidc.NewClient(ccfg)
+	if err != nil {
+		return nil, err
+	}
+
+	// SyncProviderConfig will start a goroutine to periodically synchronize the provider config.
+	// The synchronization interval is set by the expiration length of the config, and has a mininum
+	// and maximum threshold.
+	client.SyncProviderConfig(discovery)
+
+	return &OpenIDAuthenticator{client}, nil
+}
+
+// AuthenticateToken decodes and verifies a JWT using the OpenID client, then it will extract the user info
+// from the JWT claims.
+func (a *OpenIDAuthenticator) AuthenticateToken(value string) (user.Info, bool, error) {
+	jwt, err := jose.ParseJWT(value)
+	if err != nil {
+		return nil, false, err
+	}
+
+	if err := a.client.VerifyJWT(jwt); err != nil {
+		return nil, false, err
+	}
+
+	claims, err := jwt.Claims()
+	if err != nil {
+		return nil, false, err
+	}
+
+	sub, ok, err := claims.StringClaim("sub")
+	if err != nil {
+		return nil, false, err
+	}
+	if !ok {
+		return nil, false, fmt.Errorf("cannot find user subject in JWT claims")
+	}
+
+	// TODO(yifan): Add UID and Group.
```

Comment:
I think that adding the issuer would be a good thing as well as the upper layer that consume these objects (e.g. authorization) may (should actually) want to know who is asserting this identity.